pytest_asyncio/plugin.py

```diff
@@ -189,10 +189,51 @@ def pytest_runtest_setup(item):
             )
 
 
+class ClockEventLoop(asyncio.new_event_loop().__class__):
+    """
+    A custom event loop that explicitly advances time when requested. Otherwise,
+    this event loop advances time as expected.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._offset = 0
+
+    def time(self):
+        """
+        Return the time according the event loop's clock.
+
+        This time is adjusted by the stored offset that allows for advancement
+        with `advance_time`.
+        """
+        return super().time() + self._offset
+
+    async def advance_time(self, seconds):
+        '''
+        Advance time by a given offset in seconds.
+        '''
+        if seconds < 0:
+            # cannot go backwards in time, so return immediately
+            return
+
+        # advance the clock by the given offset
+        self._offset += seconds
+
+        # ensure waiting callbacks are run before advancing the clock
+        await asyncio.sleep(0, loop=self)
```

Comment:
Let me take a closer look at these sleep calls. I recall as I was writing it that I had difficulty achieving a predictable number of loop iterations to get the test to react as expected. That's how I settled on sleeping 3 times. It is also the reason I ended up using `_run_once` in the earlier version and not using a coroutine function because it's functionality was more clear (even though it was a private method call).